include/cuco/detail/trie/bit_vector/bit_vector_ref.cuh

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <cuco/operator.hpp>
+
+namespace cuco {
+namespace experimental {
+
+struct Rank;
+
+/**
+ * @brief Device non-owning "ref" type that can be used in device code to perform arbitrary
+ * operations defined in `include/cuco/operator.hpp`
+ *
+ * @tparam StorageRef Storage ref type
+ * @tparam Operators Device operator options defined in `include/cuco/operator.hpp`
+ */
+template <typename StorageRef, typename... Operators>
+class bit_vector_ref
+  : public detail::operator_impl<Operators, bit_vector_ref<StorageRef, Operators...>>... {
+ public:
+  using storage_ref_type = StorageRef;  ///< Type of storage ref
+
+  /**
+   * @brief Constructs bit_vector_ref.
+   *
+   * @param words_ref Non-owning ref of words slot storage
+   * @param ranks_ref Non-owning ref of ranks slot storage
+   * @param selects_ref Non-owning ref of selects slot storage
+   * @param ranks0_ref Non-owning ref of ranks0 slot storage
+   * @param selects0_ref Non-owning ref of selects0 slot storage
+   */
+  __host__ __device__ explicit constexpr bit_vector_ref(storage_ref_type words_ref,
+                                                        storage_ref_type ranks_ref,
+                                                        storage_ref_type selects_ref,
+                                                        storage_ref_type ranks0_ref,
+                                                        storage_ref_type selects0_ref) noexcept;
+
```

Comment:
Passing each storage one-by-one is suboptimal, since they might be incompatible (e.g. different lengths). Since this is not user-facing code, it's not a huge problem, but consider wrapping these in a single struct which internally checks for the length requirement.

Reply:
agreed. by using a struct, ranks and words can use different storage types.

we can then delete rank_union struct, whose sole purpose is to be able to use same storage type for words and ranks.

Reply:
relevant changes in c5ec25452ea9a37d355c2eaca787574bea78ec01